### ISSUE TITLE
Shrink a dick every day a bit if there was no growth last week

### ISSUE THREAD

**Issue by kozalosev** [kozalosev/DickGrowerBot#15]

**Comment by ryukaizen** [comment]
Hahahahaha this is funny as hell!

**Comment by kozalosev** [comment]
## Implementation plan

### Overview

A daily job (midnight UTC) shrinks every dick whose owner hasn't grown it in the last 7 days, records each shrink as an event, notifies affected group chats, and exposes recent events via a new inline command (so inline-only chats — which the bot can't message proactively — can still see what happened).

**Decisions:**
- **Amount:** percentage of current length per day (configurable ratio), rounded up, minimum 1 cm so small dicks still make progress.
- **Floor:** stop at 0. Only dicks with `length > 0` are shrunk; loss is clamped so length never goes below 0.
- **Group chats:** one proactive daily message per affected chat listing victims and cm lost.
- **Inline:** new inline command listing shrink events for the last N (configurable) days.

### Key enabler

The `check_and_update_dicks_timestamp` BEFORE UPDATE trigger (current version, `migrations/8`) only raises `GD0E1` when `current_date = date(OLD.updated_at)` — never true for week-inactive rows — and it does **not** set `NEW.updated_at`. So `UPDATE Dicks SET length = length - loss` on inactive rows shrinks cleanly, is never blocked, and leaves `updated_at` untouched: a neglected dick keeps shrinking every day until the owner grows it again. No trigger changes needed.

The inactivity predicate already exists (inverted) in `src/repo/users.rs`: `updated_at <= current_timestamp - interval '1 week'`. `Dick_of_Day` (`migrations/3`) is the model for an append-only event table + `forbid_*_updates` trigger.

### Steps

**1. Migration `22_create-dick-shrinks-table.sql`**
- `Dick_shrinks(chat_id bigint REFERENCES Chats(id), uid bigint REFERENCES Users(uid), lost_length bigint NOT NULL, created_at date DEFAULT current_date)`, PK `(chat_id, uid, created_at)`.
- `forbid_dick_shrinks_updates` BEFORE UPDATE trigger mirroring `forbid_dod_updates`.
- Index `dicks_idx_updated_at ON Dicks(updated_at)` for the weekly scan.

**2. Repo `src/repo/shrinks.rs`** (new `Shrinks` repo, registered in `src/repo/mod.rs`)
- `perform_daily_shrink(ratio: Ratio) -> Vec<ShrinkEvent>` — one CTE in a transaction:
  ```sql
  WITH victims AS (
    SELECT d.uid, d.chat_id,
           LEAST(d.length, GREATEST(1, CEIL(d.length * $1)::bigint)) AS loss
    FROM Dicks d
    WHERE d.length > 0
      AND d.updated_at <= current_timestamp - interval '1 week'
  ),
  updated AS (
    UPDATE Dicks d SET length = d.length - v.loss
    FROM victims v WHERE d.uid = v.uid AND d.chat_id = v.chat_id
    RETURNING d.uid, d.chat_id, v.loss, d.length AS new_length
  ),
  logged AS (
    INSERT INTO Dick_shrinks (chat_id, uid, lost_length)
    SELECT chat_id, uid, loss FROM updated
  )
  SELECT u.uid, usr.name AS owner_name, u.loss, u.new_length,
         c.chat_id AS messageable_chat_id   -- NULL for inline-only chats
  FROM updated u JOIN Users usr USING (uid) JOIN Chats c ON c.id = u.chat_id;
  ```
- `get_recent_shrinks(chat_id: &ChatIdKind, days: u32) -> Vec<ShrinkEvent>` — resolves the current chat (real *or* inline-instance) like `Dicks::fetch_length` does, filters `created_at > current_date - days`.
- Run `cargo sqlx prepare -- --tests` to refresh `.sqlx/`.

**3. Config** (`src/config/app.rs`, `toggles.rs`)
- `FeatureToggles.daily_shrink: bool` ← `SHRINK_ENABLED` (default false).
- `AppConfig.shrink_ratio: Ratio` ← `SHRINK_RATIO` (default `0.01`).
- `AppConfig.shrink_events_days: u32` ← `SHRINK_EVENTS_DAYS` (default 7).

**4. Scheduler `src/scheduler.rs`** + spawn in `src/main.rs`
- `spawn_daily_shrink(bot, repos, config)`: guard on `daily_shrink`; loop { sleep until next UTC midnight (new `duration_till_next_day_utc()` helper next to `get_time_till_next_day_string` in `src/handlers/utils/date.rs`); `run_daily_shrink(...)`; log errors, never panic }.
- In `main.rs`, clone `bot`/`repos`/`app_config` **before** `deps!` moves them, and spawn before the webhook/polling `match` so it runs in both modes.

**5. Business logic `src/handlers/shrink.rs`** (new)
- `run_daily_shrink`: call `perform_daily_shrink`, group victims by `messageable_chat_id` (skip `None` — inline-only), and `bot.send_message(...)` an HTML summary per group chat (mirror `reply_html`, but there's no `Message` to reply to).
- `recent_shrinks_impl(...)`: build the inline events list from `get_recent_shrinks`.

**6. Inline command `src/handlers/inline.rs`**
- Add `Shrinks` to the `InlineCommand` enum + an `execute` arm → `shrink::recent_shrinks_impl`. Auto-iterated, toggle-filtered (`DISABLE_CMD_SHRINKS`), round-tripped. Additionally hide it in `inline_handler` when `daily_shrink` is off.

**7. Localization** (`locales/{en,ru,fa,zh}.yml`)
- Reuse existing `commands.grow.direction.shrunk`. Add `inline.results.titles.shrinks`, `commands.shrink.notification`, `commands.shrink.recent`, `commands.shrink.empty`. `en` is the fallback.

### Open questions / limitations
- **Broadcast language:** language is tracked per *user*, not per *chat*, so a group broadcast has no natural locale — planned to default to the `en` fallback. Storing a per-chat language (the `Announcements` table already has a `language` column as precedent) would be a larger change.
- **Non-positive dicks** (already ≤ 0 from PvP/loans) are skipped, consistent with "stop at 0."

### Verification
- Repo test for `perform_daily_shrink`: seed dicks with `updated_at` older/newer than a week and positive/zero length; assert only stale positive dicks shrink, loss = `max(1, ceil(length*ratio))` clamped to ≥0, `updated_at` unchanged, one `Dick_shrinks` row per victim, and the UPDATE isn't blocked by `GD0E1`.
- Scheduler smoke test with `SHRINK_ENABLED=true`: confirm a broadcast reaches a group chat and the inline `shrinks` command lists events.
- `cargo build` / `cargo clippy`; commit the regenerated `.sqlx/`.

**Comment by kozalosev** [comment]
### Update: broadcast language for group notifications

Since this issue was drafted, a per-chat language now exists (`chats.get_chat_language`, added in
#125), which resolves the "broadcast has no natural locale" open question for chats that have a
chat-wide language set.

**Decision on the fallback** (when a chat has *no* chat-wide language): use the most popular
preferred language among that chat's players, tallied **client-side** (it's our broadcast policy,
and the tally can feed our existing `USER_CACHE_TIME_SECS` cache). This needs a batch lookup that
user-service doesn't have yet, so I've filed it there:

- Depends on: **Kozalo-Blog/user-service#25** — add a typed `getUsers(repeated ids, optional FieldMask)` RPC.

Other stale details in the plan above, corrected during implementation planning: next migration is
**27** (not 22); there's no `src/scheduler.rs`/`src/handlers/utils/date.rs` yet (the daily job is
modeled on the `self_destruction.rs` spawn pattern, and the duration helper goes in the inline
`date` module in `utils/mod.rs`); the Dicks BEFORE UPDATE trigger already doesn't touch
`updated_at`, so no trigger change is needed; and there are 6 locale files (`en, ru, it, fa, zh, zh-TW`).
